File: Python_practice/Algorithm/algorithm_I_5_heapsort.py
#建立堆
#逐个出数
#直到堆空

def sift (list,low,high):
    temp = list[low]#堆的根结点
    i = low
    j = 2*i + 1 
    while (j <= high):
        if (j+1 <= high and list[j+1] > list[j] ):
            j += 1 
        if temp < list[j] : 
            list[i] = list[j] #如果temp不够格，那么就让现在的j上去补位，空出现在的j
            i = j
            j = 2*i + 1  
        else: 
            break
    list[i] = temp#不论如何总会出来这里接手的


def heap_sort (list):
#建立堆： 
    n = len ( list )
    for i in range((n-2)//2 ,-1 ,-1 ):#若最后一个下标 i : (i-1) / 2就是父节点 ；i = n - 1 ; 故有 
        sift (list,i,n-1) 
    #把最后堆元素提上来，然后做一次向下调整，直到堆空排序完成；
    for i in range (n-1,-1,-1):
        list[0] ,list [i] = list[i],list[0] 
        sift ( list , 0 ,i -1 ) 
    print(list)

    
list = [5,6325,168,514,987,526,256,46,145,845,145,846,15,9586,134,856,125,4215]
print(list)
heap_sort(list)

File: C/Algorithm/algorithm_I_5_heap_sort.c
#include <stdio.h>
#include <stdlib.h>

int sift(int *list, int low, int high);

int sift(int *list, int low, int high)
{
    int temp = list[low];
    int i = low;
    int j = 2 * i + 1;
    //初始化

    while (j <= high)
    {
        if (j + 1 <= high && list[j + 1] > list[j])
        {
            j += 1;
        }//选择大的子元素

        if (temp < list[j])
        {
            list[i] = list[j];
            i = j;
            j = 2 * i + 1;
        }//如果不够格，就把下一个移上去，然后继续向下寻找

        else
        {
            break;//够格的话就要跳出去
        }
    }
    list[i] = temp;//不论是else中break还是while不符合跳出，最后都要把temp放回去
    return 0;
}

int heap_sort(int *list, int n);
int heap_sort(int *list, int n)
{
    //建立堆:list数组；n是数组长度
    for (int i = (n - 2) / 2; i >= 0; i--)
    {
        sift(list, i, n - 1);
    }
    //逐个出数
    for (int i = n - 1; i >= 0; i--)
    {
        int tmp = list[0];
        list[0] = list[i];
        list[i] = tmp;//把最后堆的最后一个数提上来，堆顶元素放下去，然后做向下调整
        sift(list,0,i-1);//注意这里是i-1，因为最后一个元素刚刚是堆顶拿下来的，不是堆里的元素
    }
    return 0;
}

int main()
{
    int a[10] = {9, 8, 7, 658463, 73626737, 3, 2, 1,1,78316};
    int j = 0;
    while (a[j] != 0)
    {
        j++;
    }
    for (int i = 0; i < j; i++)
    {
        printf("%d ", a[i]);
    }
    printf("\n");
    heap_sort(a, j);
    for (int i = 0; i < j; i++)
    {
        printf("%d ", a[i]);
    }
    return 0;
}